ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Impl/Pure.hs

```diff
@@ -0,0 +1,321 @@
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | Pure side of the Mempool implementation.
+--
+-- Operations are performed in a pure style returning data types that model
+-- the control flow through the operation and can then be interpreted to perform
+-- the actual STM/IO operations.
+
+module Ouroboros.Consensus.Mempool.Impl.Pure (
+    -- * Mempool
+    pureGetSnapshotFor
+  , pureRemoveTxs
+  , pureSyncWithLedger
+  , pureTryAddTxs
+  , runRemoveTxs
+  , runSyncWithLedger
+  , runTryAddTxs
+    -- * MempoolSnapshot
+  , implSnapshotFromIS
+  ) where
+
+import           Control.Exception (assert)
+import           Data.Maybe (isJust, isNothing)
+import qualified Data.Set as Set
+import           Data.Word (Word32)
+
+import           Control.Tracer
+
+import           Ouroboros.Consensus.HeaderValidation
+import           Ouroboros.Consensus.Ledger.Abstract
+import           Ouroboros.Consensus.Ledger.SupportsMempool
+import           Ouroboros.Consensus.Mempool.API
+import           Ouroboros.Consensus.Mempool.Impl.Types
+import           Ouroboros.Consensus.Mempool.TxSeq (TicketNo, TxTicket (..),
+                     zeroTicketNo)
+import qualified Ouroboros.Consensus.Mempool.TxSeq as TxSeq
+import           Ouroboros.Consensus.Util.IOLike
+
+
+{-------------------------------------------------------------------------------
+  Mempool Implementation
+-------------------------------------------------------------------------------}
+
+-- | Data constructor containing the control flow for 'tryAddTxs'. Can be
+-- interpreted in an IOLike monad with 'runTryAddTxs'.
+data TryAddTxs blk =
+  Done [GenTx blk]
```

Comment:
Style: `Done` should be indented the same as `Trace`.

Reply:
Indeed, fixed. (resolving)